Add local refactoring code actions based on swift-syntax

DougGregor opened the pull request:
Introduce new local refactoring code actions based on the Swift syntax tree, without going through SourceKit. 

This change includes a number of new refactoring code actions. Most of them adapt the syntax refactorings from the [`SwiftRefactor` module of swift-syntax](https://github.com/apple/swift-syntax/tree/main/Sources/SwiftRefactor):
* Add digit separators to an integer literal, e.g., `100000` -> `1_000_000`. 
* Remove digit separators from an integer literal, e.g., `1_000_000` -> `1000000`. (from SwiftRefactor)
* Format a raw string literal, e.g., `"Hello \#(world)"` ->  `##"Hello \#(world)"##`
* Migrate to new if let syntax, e.g., `if let x = x { ... }` -> `if let x { ... }`
* Replace opaque parameters with generic parameters, e.g., `func f(p: some P)` --> `func f<T1: P>(p: T1)`.

This is generally easy to do, requiring one conformance to provide a name for the refactoring:

    extension AddSeparatorsToIntegerLiteral: SyntaxRefactoringCodeActionProvider {
      public static var title: String { "Add digit separators" }
    }

There are also a few new refactoring code actions implemented here. These could be sunk down into SwiftRefactor to be made more widely available, or stay here if they're too IDE-centric to live in swift-refactor:

* Convert an integer literal from one base to another (e.g., `172387` -> `0x2a163`). 
* Convert JSON into a Codable struct.
* Apply Demorgan's law, e.g., `!(a || b)` --> `!a && !b`.

There is a small amount of overlap between the refactoring code actions added here and [those SourceKit provides](https://github.com/apple/swift/blob/main/tools/swift-refactor/swift-refactor.cpp). For example, SourceKit provides a "Simplify long number literal" operation that is similar to "Add digit separators". We have several directions we could go with those:
1. Don't add the new refactoring code actions if they conflict. This would be sad.
2. Filter out the SourceKit ones so the user doesn't see them. We can keep the list in SourceKit-LSP.
3. Remove the redundant actions from SourceKit. Clients of SourceKit (not -LSP) would lose access to them.
4. Replace the SourceKit implementation with the ones from SwiftRefactor to server SourceKit clients, and filter them out (like option 2) for LSP clients.

Most of the code here is actually from @CodaFi, and I've given it a light dusting to bring it up to modern times.

DougGregor commented:
@swift-ci please test

CodaFi commented on Sources/SourceKitLSP/Swift/CodeActions/Demorgan.swift:
````diff
@@ -0,0 +1,200 @@
+/// ## After
+/// ```swift
+///
+/// ```
````
Heh the fact I didn't finish documenting this likely means it may not be ready for prime time. I dunno, tho, the implementation looks mostly fine.

CodaFi reviewed (approved):
❤️

CodaFi commented on Sources/SourceKitLSP/Swift/CodeActions/ConvertJSONToCodableStruct.swift:
```diff
@@ -0,0 +1,290 @@
+    }
+
+    var text = ""
+    switch unexpected {
+    case let .closure(closure):
+      closure.write(to: &text)
+    case let .tail(closure, unexpected):
+      closure.write(to: &text)
+      unexpected.write(to: &text)
+    }
```
```suggestion
    switch unexpected {
    case let .closure(closure):
      closure.trimmed.write(to: &text)
    case let .tail(closure, unexpected):
      closure.trimmed.write(to: &text)
      unexpected.write(to: &text)
    }
```

CodaFi replied:
Otherwise we'll pick up comments!

// Foo Bar Baz

{
 "abc": "xyz"
}

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test windows

DougGregor commented:
@swift-ci please test Windows

fwcd commented:
Love the JSON-to-Codable conversion, that looks super useful!

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test Windows

DougGregor commented:
@swift-ci please test Windows

ahoppen commented on Sources/SourceKitLSP/Swift/CodeActions/ConvertJSONToCodableStruct.swift:
```diff
@@ -0,0 +1,287 @@
+  ) -> [String] {
+    var members = [String]()
+
+    var uniquer = Set<String>()
+    var nestedTypes = [(String, Dictionary<String, Any>)]()
```
Why can’t we use a `[String: [String: Any]]` here instead of `uniquer` and an array containing a tuple? We do sort the keys from the `jsonDictionary` anyway, so I think there’s no order we need to preserver (if we sort by the dictionary keys in the end).

ahoppen commented on Sources/SourceKitLSP/Swift/CodeActions/ConvertJSONToCodableStruct.swift:
```diff
@@ -0,0 +1,287 @@
+      nestedTypes.append((name, object))
+    }
+
+    for key in jsonDictionary.keys.sorted() {
+      guard let value = jsonDictionary[key] else {
+        continue
+      }
```
I think this can be simplified

```suggestion
    for (key, value) in jsonDictionary.sorted(\.key) {
```

ahoppen commented on Sources/SourceKitLSP/Swift/CodeActions/ConvertJSONToCodableStruct.swift:
```diff
@@ -0,0 +1,287 @@
+  }
+
+  private static func jsonType(of value: Any, suggestion name: String) -> JSONType {
+    switch Swift.type(of: value) {
+    case is NSString.Type:
+      return .string
```
Do we need to use `Swift.type(of:)` here? Can’t we just use

```suggestion
    switch value {
    case is NSString.self:
      return .string
```

ahoppen commented on Sources/SourceKitLSP/Swift/CodeActions/ConvertJSONToCodableStruct.swift:
```diff
@@ -0,0 +1,287 @@
+    case is NSNumber.Type:
+      return .double
+    case is NSArray.Type:
+      guard let firstValue = (value as! [Any]).first else {
+        return .array(.null)
+      }
+      let innerType = self.jsonType(of: firstValue, suggestion: name)
+      return .array(innerType)
```
Hmm, I think it’s unfortunate that we don’t allow merging the keys from multiple objects in the array. I don’t think it’s uncommon for JSON to omit keys for null values. We should be able to take the union of all the keys in the array. And I think it would also be good to infer optionality of a value eg. from

```json
[
  {
    "value": 1
  },
  {
    "value": null
  }
]
```

ahoppen commented on Sources/SourceKitLSP/Swift/CodeActions/ConvertJSONToCodableStruct.swift:
```diff
@@ -0,0 +1,287 @@
+    var values: [Any] = self
+    for i in 0..<depth {
+      if i + 1 == depth {
+        return values[0] as? Dictionary<String, Any>
```
Do we know that `self` is not empty here?

ahoppen commented on Sources/SourceKitLSP/Swift/CodeActions/ConvertJSONToCodableStruct.swift:
```diff
@@ -0,0 +1,287 @@
+}
+
+extension Array where Element == Any {
+  fileprivate func unwrap(_ depth: Int) -> Dictionary<String, Any>? {
```
Could you add a comment of what `unwrap` does?

ahoppen commented on Sources/SourceKitLSP/Swift/CodeActions/ConvertJSONToCodableStruct.swift:
```diff
@@ -0,0 +1,287 @@
+    }
+
+    if !nestedTypes.isEmpty {
+      members.append("")
```
Having to append an empty string to members to (presumably) bet an empty line between the members seems wrong. I think we should add an empty line between all the members instead.

ahoppen commented on Sources/SourceKitLSP/Swift/CodeActions/ConvertJSONToCodableStruct.swift:
```diff
@@ -0,0 +1,287 @@
+    case null
+    case object(String)
+
+    var outermostObject: (Int, String)? {
```
Instead of returning `unwraps` here and then calling `unwrap` with it, wouldn’t it be easier to just return the name and `JSONType.object` here?

DougGregor commented on Sources/LanguageServerProtocol/Requests/CodeActionRequest.swift:
```diff
@@ -10,7 +10,7 @@
 //===----------------------------------------------------------------------===//
 
-public typealias CodeActionProvider = (CodeActionRequest) async throws -> [CodeAction]
+public typealias CodeActionRequestProvider = (CodeActionRequest) async throws -> [CodeAction]
```
It'll go back to the old name and rename the protocol

DougGregor commented on Sources/SourceKitLSP/Swift/CodeActions/CodeActionProvider.swift:
```diff
@@ -0,0 +1,75 @@
+import SwiftRefactor
+import SwiftSyntax
+
+public protocol CodeActionProvider {
```
Okay!

DougGregor commented on Sources/SourceKitLSP/Swift/CodeActions/CodeActionProvider.swift:
```diff
@@ -0,0 +1,75 @@
+
+public protocol CodeActionProvider {
+  static var kind: CodeActionKind { get }
+  static func provideAssistance(in scope: CodeActionScope) -> [ProvidedAction]
```
We can remove `ProvidedAction` in favor of `CodeAction`; I've done so and renamed to `codeActions(in:)`

DougGregor commented on Sources/SourceKitLSP/Swift/CodeActions/CodeActionProvider.swift:
```diff
@@ -0,0 +1,75 @@
+    self.parameters = parameters
+    self.file = tree
+
+    let start = self.snapshot.utf8Offset(of: self.parameters.range.lowerBound) ?? 0
+    let end = self.snapshot.utf8Offset(of: self.parameters.range.upperBound) ?? start
```
Okay!

DougGregor commented on Sources/SourceKitLSP/Swift/CodeActions/CodeActionProvider.swift:
```diff
@@ -0,0 +1,75 @@
+}
+
+extension SyntaxProtocol {
+  var textRange: ByteSourceRange {
+    return ByteSourceRange(
+      offset: self.positionAfterSkippingLeadingTrivia.utf8Offset,
+      length: self.trimmedLength.utf8Length
+    )
+  }
```
Yeah, we don't even need this. A `Range<AbsolutePosition>` works for our purposes.

DougGregor commented on Sources/SourceKitLSP/Swift/CodeActions/ConvertIntegerLiteral.swift:
```diff
@@ -0,0 +1,67 @@
+import SwiftRefactor
+import SwiftSyntax
+
+// TODO: Make the type IntegerLiteralExprSyntax.Radix conform to CaseEnumerable
+// in swift-syntax.
```
grumble grumble dependent PRs grumble grumble

DougGregor commented on Sources/SourceKitLSP/Swift/CodeActions/ConvertIntegerLiteral.swift:
```diff
@@ -0,0 +1,67 @@
+  static var allCases: [Self] = [.binary, .octal, .decimal, .hex]
+}
+
+public struct ConvertIntegerLiteral: CodeActionProvider {
```
We could do this, yes.

DougGregor commented on Sources/SourceKitLSP/Swift/CodeActions/ConvertIntegerLiteral.swift:
```diff
@@ -0,0 +1,67 @@
+}
+
+public struct ConvertIntegerLiteral: CodeActionProvider {
+  public static var kind: CodeActionKind { .refactorInline }
```
I don't have a good sense of how these manifest. Does this go for all of the refactoring we're creating?

ahoppen replied:
I don’t actually know either (I guess also depends on the IDE) but it’s good practice to choose the closest possible case.

DougGregor commented on Sources/SourceKitLSP/Swift/CodeActions/EditBuilder.swift:
```diff
@@ -0,0 +1,76 @@
+  }
+}
+
+public struct ProvidedAction {
+  public var title: String
+  public var edits: [BuildableWorkspaceEdit]
+
+  public init(title: String, edits: [BuildableWorkspaceEdit]) {
+    self.title = title
+    self.edits = edits
+  }
+
+  public init(title: String, @WorkspaceEditBuilder edit: () -> [BuildableWorkspaceEdit]) {
+    self.title = title
+    self.edits = edit()
+  }
+}
```
I'm removing it entirely

DougGregor commented on Sources/SourceKitLSP/Swift/SwiftLanguageService.swift:
```diff
@@ -729,6 +732,22 @@ extension SwiftLanguageService {
     }.flatMap { $0 }
   }
 
+  func retrieveLocalCodeActions(_ params: CodeActionRequest) async throws -> [CodeAction] {
```
I'm going with `retrieveSyntaxCodeActions`

DougGregor commented on Sources/SourceKitLSP/Swift/SwiftLanguageService.swift:
```diff
@@ -729,6 +732,22 @@ extension SwiftLanguageService {
+        ])
+        return CodeAction(title: $0.title, kind: provider.kind, edit: edit)
+      }
+    }.flatMap { $0 }
```
Sure!

DougGregor commented on Sources/SourceKitLSP/Swift/SwiftLanguageService.swift:
```diff
@@ -700,7 +700,8 @@ extension SwiftLanguageService {
   public func codeAction(_ req: CodeActionRequest) async throws -> CodeActionRequestResponse? {
-    let providersAndKinds: [(provider: CodeActionProvider, kind: CodeActionKind)] = [
+    let providersAndKinds: [(provider: CodeActionRequestProvider, kind: CodeActionKind)] = [
+      (retrieveLocalCodeActions, .refactorRewrite),
```
Yeah, I want to do this a different way. Thanks

DougGregor commented on Sources/SourceKitLSP/Swift/CodeActions/CodeActionProvider.swift:
```diff
@@ -0,0 +1,75 @@
+  public var snapshot: DocumentSnapshot
+  public var parameters: CodeActionRequest
+  public var file: SourceFileSyntax
+  public var range: ByteSourceRange
```
I've switched over to Range<AbsolutePosition>

DougGregor commented on Sources/SourceKitLSP/Swift/CodeActions/ConvertIntegerLiteral.swift:
```diff
@@ -0,0 +1,67 @@
+  public static func provideAssistance(in scope: CodeActionScope) -> [ProvidedAction] {
+    guard
+      let token = scope.file.token(at: scope.range.offset),
+      let lit = token.parent?.as(IntegerLiteralExprSyntax.self),
```
Eh, sure

DougGregor commented on Sources/SourceKitLSP/Swift/CodeActions/EditBuilder.swift:
```diff
@@ -0,0 +1,76 @@
+import LanguageServerProtocol
+import SwiftSyntax
+
+@resultBuilder
+struct WorkspaceEditBuilder {
+  static func buildBlock(_ components: BuildableWorkspaceEdit...) -> [BuildableWorkspaceEdit] {
+    return components
+  }
+}
```
I agree, done

DougGregor commented on Sources/SourceKitLSP/Swift/CodeActions/EditBuilder.swift:
```diff
@@ -0,0 +1,76 @@
+      scope.snapshot.positionOf(utf8Offset: self.old.positionAfterSkippingLeadingTrivia.utf8Offset)
+      ?? Position(line: 0, utf16index: 0)
+    let upper = scope.snapshot.positionOf(utf8Offset: oldRange.endOffset) ?? Position(line: 0, utf16index: 0)
+    return TextEdit(range: Range(uncheckedBounds: (lower: lower, upper: upper)), newText: "\(self.new)")
```
I'm removing this code

DougGregor commented on Sources/SourceKitLSP/Swift/CodeActions/SyntaxRefactoringCodeActionProvider.swift:
```diff
@@ -0,0 +1,66 @@
+}
+
+extension FormatRawStringLiteral: SyntaxRefactoringCodeActionProvider {
+  public static var title: String { "Format raw string literal" }
```
Sure!

DougGregor commented on Sources/SourceKitLSP/Swift/CodeActions/SyntaxRefactoringCodeActionProvider.swift:
```diff
@@ -0,0 +1,66 @@
+}
+
+extension MigrateToNewIfLetSyntax: SyntaxRefactoringCodeActionProvider {
+  public static var title: String { "Migrate to 'if-let' Syntax" }
```
Okay!

DougGregor commented on Sources/SourceKitLSP/Swift/CodeActions/SyntaxRefactoringCodeActionProvider.swift:
```diff
@@ -0,0 +1,66 @@
+}
+
+extension OpaqueParameterToGeneric: SyntaxRefactoringCodeActionProvider {
+  public static var title: String { "Expand opaque parameters to generic parameters" }
```
Sure

DougGregor commented:
I'm tackling the first part of this, exposing swift-syntax-based actions, via the PR at https://github.com/apple/sourcekit-lsp/pull/1179. It drops Demorgans and the JSON-to-structs refactoring actions and addresses most of the feedback here. (Not all of it, yet)

DougGregor commented:
Everything but DeMorgan's has merged via other pull requests, so I'm going to close this out. DeMorgan's needs a some work that I'm not up for just now.
